### PR TITLE
Allow customiztion before command executes.

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

As an end user, sometimes I want to provide some extra arguments to the commands to be executed by `exemplar`, for example, in `gradle/gradle` we want to add `-Dorg.gradle.internal.plugins.portal.url.override=<OurOwnRepositoryMirror>`. Currently it's not supported by `exemplar` - `exemplar` will only run the commands declared in conf files, there's no way to customize them.

This PR adds an annotation `@CommandCustomizers` to allow users to customize their `Command`s before execution. A typical usage is:

```
@RunWith(GradleSamplesRunner.class)
@SamplesRoot("src/test/samples/customization")
@CommandCustomizers(ExtraTaskRequestCustomizer.class)
public class CustomizedExecutionIntegrationTest {
}
```

```
public class ExtraTaskRequestCustomizer implements CommandCustomizer {
    @Override
    public Command customize(Command command) {
        List<String> args = new ArrayList<>(command.getArgs());
        args.add("extraArgument");
        return CommandBuilder.fromCommand(command).setArgs(args).build();
    }
}
```

@eriwen Can you please take a look?

### CODE REVIEW

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 1]
```diff
@@ -19,4 +21,6 @@ object Libraries {
     val SLF4J = "org.slf4j:slf4j-simple:${Versions.SLF4J}"
     val SPOCK_CORE = "org.spockframework:spock-core:${Versions.SPOCK_CORE}"
     val TYPESAFE_CONFIG = "com.typesafe:config:${Versions.TYPESAFE_CONFIG}"
+    val CGLIB = "cglib:cglib-nodep:${Versions.CGLIB}"
+    val OBJENESIS = "org.objenesis:objenesis:${Versions.OBJENESIS}"
```
Comment (blindpirate) [inline buildSrc/src/main/kotlin/Dependencies.kt]: These two libraries are necessary for spock `Mock()`.

**Comment by eriwen** [comment]
Hey Bo, could you please describe the end user needs here?

**Comment by blindpirate** [comment]
@eriwen Sure, I explained a little above. Please let me know if I didn't make myself clear, thanks!

**Comment by blindpirate** [comment]
Also, it seems that we don't have any CI configuration for this repository, do we?

**Comment by blindpirate** [comment]
@eriwen Could you please take a look? We've been bitten by network issues in sample tests for a long time....

File: sample-check/src/main/java/org/gradle/samples/test/customizer/CommandCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.gradle.samples.test.customizer;
+
+import org.gradle.samples.model.Command;
+
+/**
+ * Customize the command before the execution runs.
+ */
+
+public interface CommandCustomizer {
```
Comment (eriwen) [inline sample-check/src/main/java/org/gradle/samples/test/customizer/CommandCustomizer.java]: I prefer the term "Modifier" here

**Comment by eriwen** [comment]
@blindpirate I adjusted this PR slightly and merged with https://github.com/gradle/exemplar/commit/a1e9ad658d5e01406b35697cbcf6568f53a4fb73:

Allow Samples to be modified instead of just Commands. This allows folks to prepend/append commands to all samples in addition to modifying them all.
I also changed up the terminology and moved things around a bit. The integration tests still pass, so I imagine this'll work just fine for Gradle. 

I've been trying to figure out why we get "too long path" errors on Windows in the Gradle build using Exemplar v0.5+. I'm working on fixing [this problem](https://gradle.slack.com/archives/C09LZKUCB/p1535651003000100) but happy to get your eyes on it if you need the fix sooner. 

Please reopen this PR or an issue if what I've provided doesn't actually do what you need for Gradle.

**Pull request closed without merging**